File: 算法/quicksort/qs.c
#include <stdio.h>

void pprint(int *a)
{
    for (int i = 0; i < 10; i++) {
        //printf("%d:%d   ", i, a[i]);
        printf("%d   ", a[i]);
    }
    printf(" \n");
}

void qsort(int *a, int left, int right)
{
    if (left >= right) {
        return ;
    }

    int i = left;
    int j = right;
    int val = a[left];

    while (i < j) {
        while (i < j && val <= a[j]) { // 如果i所在值比右侧j所在位置的值小 不用动j指针往前挪
            j--; // 继续往左找
        }

        // 如果找到比val值小的 就换到i 的位置
        a[i] = a[j];
        printf("1 i:%d   j:%d \n", i, j);

        while (i < j && val >= a[i])
        {
            i++;
        }
        a[j] = a[i];
        printf("2 i:%d   j:%d \n", i, j);
    }
    a[i] = val;

    pprint(a);

    qsort(a, left, i - 1);
    qsort(a, i + 1, right);
}

int main()
{
    int arr[10] = {6,2,3,10,1,4,7,8,9,5};
    qsort(arr, 0, 9);
    pprint(arr);
    return 0;
}

File: 算法/quicksort/qs.php
<?php

$arr = [6,2,3,10,1,4,7,8,9,5];

function qs(&$arr, $left, $right)
{
    if ($left >= $right) {
        return ;
    }

    $i = $left;
    $j = $right;
    $val = $arr[$left];

    while ($i < $j) {
        while ($i < $j && $val < $arr[$j])
        {
            $j--; // 往左走
        }
        $arr[$i] = $arr[$j];

        while ($i < $j && $val >= $arr[$i])
        {
            $i++; // 往右走
        }
        $arr[$j] = $arr[$i];
    }
    $arr[$i] = $val;
    qs($arr, $left, $i -1 );
    qs($arr, $i + 1, $right );
}
//qs($arr, 0 , 9);

function maopao($arr)
{
    $cnt = count($arr);
    for ($i = 0; $i < $cnt; $i++) {
        for ($j = $cnt - 1; $j > $i; $j--)
        {
            if ($arr[$j - 1] > $arr[$j]) {
                $x = $arr[$j];
                $arr[$j] = $arr[$j - 1];
                $arr[$j - 1] = $x;
            }
        }
    }
    return $arr;
}
$arr = maopao($arr);
var_dump($arr);
//var_dump($arr);
